### PR TITLE
Feat: register new candidates to a race, add a race to an election, r…

### PR DESCRIPTION
**Pull request by odellmac4**

### CODE REVIEW

File: spec/candidate_spec.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
+require './lib/candidate'
+
+RSpec.describe Candidate do
+    before(:each) do
+      @diana = Candidate.new({name: "Diana D", party: :democrat})
+  
+    end
```
Comment (KatBrandt) [inline spec/candidate_spec.rb:7]: Nit pick: space after this end.

File: spec/candidate_spec.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+            @diana.vote_for!
+            
+            expect(@diana.votes).to eq (3)
+
+            @diana.vote_for!
+
+            expect(@diana.votes).to eq (4)
+        end
+    end
+end
```
Comment (KatBrandt) [inline spec/candidate_spec.rb:39]: Great, tests all look good here.

File: lib/candidate.rb [hunk 3]
```diff
@@ -1,2 +1,16 @@
...
+    def initialize(info)
+        @name = info[:name]
+        @party = info[:party]
+        @votes = 0
+    end
+
+    def vote_for!
+        @votes += 1
+    end
 end
```
Comment (KatBrandt) [inline lib/candidate.rb:16]: Yup, this class looks good.

File: lib/race.rb [hunk 4]
```diff
@@ -1,2 +1,16 @@
...
+
+    def initialize(office)
+        @office = office
+        @candidates = []
+    end
+
+    def register_candidate!(info)
+        new_candidate = Candidate.new(info)
+        @candidates << new_candidate
+        return new_candidate
```
Comment (KatBrandt) [inline lib/race.rb:14]: Great job recognizing that we did want a specific return value on this command method. Remember that you don't need the `return` keyword as ruby uses implicit return.

File: spec/race_spec.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+            expect(candidate1.name).to eq ("Diana D")
+            expect(candidate1.party).to eq (:democrat)
+
+            candidate2 = @race.register_candidate!({name: "Roberto R", party: :republican})
+
+            expect(candidate2.class).to eq (Candidate)
+            
+            expect(@race.candidates).to eq ([candidate1 , candidate2])
+        end
+    end
```
Comment (KatBrandt) [inline spec/race_spec.rb:36]: Good test. I like that you tested that multiple objects could be added to the candidates array.

File: lib/election.rb [hunk 6]
```diff
@@ -1,2 +1,31 @@
...
+        races << race
+    end
+
+    def candidates
+        all_candidates = []
+        @races.each do |race|
+            all_candidates << race.candidates
+        end
+        all_candidates.flatten
+
```
Comment (KatBrandt) [inline lib/election.rb:22]: Yup, this method works great. You can also try using .map or .flat_map as well instead of using .each if you want to try those out.

File: lib/election.rb [hunk 7]
```diff
@@ -1,2 +1,31 @@
...
+        all_candidates.flatten
+        
+    end
+
+    def vote_counts
+        candidates.each do |candidate|
+            @candidate_votes[candidate.name] = candidate.votes
+        end
+        @candidate_votes
+    end
```
Comment (KatBrandt) [inline lib/election.rb:30]: This method works and will give the desired result. I don't know that you need to create an instance variable for this, you could instead keep this calculation contained within this method so anytime you call `vote_counts` it will recalculate so then no need to store anything in an instance variable. The other option is to store this info in the instance variable, but have the instance variable updated whenever a candidate a race is added, then you'd just have an attr_reader for the :vote_counts and have that as the instance variable. 
Basically we want to think about, do I want this data to be recalculated every time? If so, then I only need a method. If I want it to be stored for future use, then I should build it in a way where the data is not recalculated every time, but instead is added to the instance variable upon some other action (like whenever we add a race). 
I hope this makes sense, happy to talk more about it if it doesn't. 
Long story short though, this would work just fine for an IC, but for future projects, you'd want to restructure this a bit.

**Review by KatBrandt — COMMENTED** [review]
Overall really great job! It seems like you understand all the concepts being tested here. Just a heads up, the week 5 IC will be quite a bit more challenging than this week 4 IC. You got this!

File: lib/election.rb (hunk 7 above)
Comment (odellmac4) [reply lib/election.rb:30]: Yes! This makes complete sense. Maybe I can look up more examples of this just to familiarize myself with this concept more. Seems like instance variables may be a little fuzzy still.
